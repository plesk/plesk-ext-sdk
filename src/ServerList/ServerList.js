// Copyright 1999-2018. Plesk International GmbH. All rights reserved.

import { Component, List, Pagination, PropTypes } from '@plesk/ui-library';
import { withApi } from '../Api';
import { withConfig } from '../Config';
import { withStatusMessages } from '../StatusMessages';

export class ServerList extends Component {
    static propTypes = {
        config: PropTypes.shape({
            baseUrl: PropTypes.string.isRequired,
        }).isRequired,
        api: PropTypes.shape({
            get: PropTypes.func.isRequired,
        }).isRequired,
        statusMessages: PropTypes.shape({
            add: PropTypes.func.isRequired,
        }).isRequired,
        action: PropTypes.string.isRequired,
        columns: PropTypes.arrayOf(PropTypes.shape({
            key: PropTypes.string.isRequired,
            title: PropTypes.node,
            render: PropTypes.func,
            sortable: PropTypes.bool,
        })).isRequired,
        defaultPageSize: PropTypes.number,
        defaultSortColumn: PropTypes.string,
        defaultSortDirection: PropTypes.oneOf(['ASC', 'DESC']),
        listProps: PropTypes.shape(),
        paginationProps: PropTypes.shape(),
    };

    static defaultProps = {
        listProps: {},
        paginationProps: {},
        defaultPageSize: 25,
        defaultSortColumn: null,
        defaultSortDirection: 'ASC',
    };

    state = {
        pageSize: this.props.defaultPageSize,
        pageNumber: 1,
        totalPages: 0,
        sortColumn: this.props.defaultSortColumn,
        sortDirection: this.props.defaultSortDirection,
        data: [],
    };

    componentDidMount() {
        this.fetchItems();
    }

    fetchItems = () => {
        const { config, action, api, statusMessages } = this.props;
        const { pageSize, pageNumber, sortColumn, sortDirection } = this.state;
        const params = {
            pageSize,
            pageNumber,
        };

        if (sortColumn) {
            params.sortColumn = sortColumn;
            params.sortDirection = sortDirection;
        }

        api.get(`${config.baseUrl}${action}`, params).then(({ status, data, totalItems, errors }) => {
            if (status === 'error') {
                errors.forEach(message => statusMessages.add({ intent: 'danger', message }));
                this.setState({
                    pageNumber: 1,
                    totalPages: 0,
                });
            } else {
                this.setState({
                    data,
                    totalPages: Math.ceil(totalItems / pageSize),
                });
            }
        });
    };

    handleSelectPage = pageNumber => {
        const { paginationProps } = this.props;
        const { onSelect } = paginationProps;

        if (typeof onSelect === 'function' && onSelect(pageNumber) === false) {
            return;
        }

        this.setState({
            pageNumber,
        }, () => {
            this.fetchItems();
        });
    };

    handleSortChange = params => {
        const { listProps } = this.props;
        const { onSortChange } = listProps;

        if (typeof onSortChange === 'function' && onSortChange(params) === false) {
            return;
        }

        this.setState(params, () => this.fetchItems());
    };

    renderPagination = () => {
        const { paginationProps } = this.props;
        const { totalPages, pageNumber } = this.state;

        if (totalPages < 2) {
            return null;
        }

        return (
            <Pagination
                {...paginationProps}
                total={totalPages}
                current={pageNumber}
                onSelect={this.handleSelectPage}
            />
        );
    };

    render() {
        const {
            config,
            action,
            api,
            statusMessages,
            columns,
            defaultPageSize,
            defaultSortColumn,
            defaultSortDirection,
            listProps,
            paginationProps,
            ...props
        } = this.props;
        const { data, sortColumn, sortDirection } = this.state;

        return (
            <div {...props}>
                <List
                    {...listProps}
                    columns={columns}
                    sortColumn={sortColumn}
                    sortDirection={sortDirection}
                    onSortChange={this.handleSortChange}
                    data={data}
                />
                {this.renderPagination()}
            </div>
        );
    }
}

export default withStatusMessages(withApi(withConfig(ServerList)));
