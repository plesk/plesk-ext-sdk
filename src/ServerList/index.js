// Copyright 1999-2025. WebPros International GmbH. All rights reserved.

export { default as ServerList } from './ServerList';
