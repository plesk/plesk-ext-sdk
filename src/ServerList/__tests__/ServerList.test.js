// Copyright 1999-2018. Plesk International GmbH. All rights reserved.

import { Pagination, List } from '@plesk/ui-library';
import { ServerList } from '../ServerList';
import { shallow } from 'enzyme';

const generateData = length => [...Array(length)].map((_, index) => ({ key: String(index), column1: index }));

describe('ServerList', () => {
    let props;

    beforeEach(() => {
        props = {
            config: {
                baseUrl: '',
            },
            api: {
                get: jest.fn(),
            },
            statusMessages: { add: jest.fn() },
            columns: [{
                key: 'column1',
                title: 'Column1',
            }],
            action: '/api/test',
        };
    });

    test('renders correctly', () => {
        const data = generateData(5);
        const promise = Promise.resolve({ data, totalItems: data.length });
        props.api.get.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerList
                {...props}
            />,
        );

        expect.assertions(4);

        return promise.then(() => {
            expect(props.api.get).toHaveBeenCalledWith('/api/test', { pageSize: 25, pageNumber: 1 });
            expect(wrapper.state()).toMatchObject({
                pageNumber: 1,
                totalPages: 1,
                sortColumn: null,
                sortDirection: 'ASC',
            });
            expect(wrapper.state().data).toStrictEqual(data);
            expect(wrapper).toMatchSnapshot();
        });
    });

    test('renders pagination correctly', () => {
        const data = generateData(60);
        const promise = Promise.resolve({ data: data.slice(0, 25), totalItems: data.length });
        props.api.get.mockReturnValueOnce(promise);
        props.api.get.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerList
                {...props}
            />,
        );

        expect.assertions(5);

        return promise
            .then(() => {
                expect(props.api.get).toHaveBeenCalledWith('/api/test', { pageSize: 25, pageNumber: 1 });
                expect(wrapper.state()).toMatchObject({
                    pageNumber: 1,
                    totalPages: 3,
                });
                expect(wrapper).toMatchSnapshot();

                wrapper.find(Pagination).simulate('select', 2);

                expect(props.api.get).toHaveBeenCalledWith('/api/test', { pageSize: 25, pageNumber: 2 });
                expect(wrapper.state()).toMatchObject({
                    pageNumber: 2,
                    totalPages: 3,
                });
            });
    });

    test('renders sort correctly', () => {
        const data = generateData(60);
        const promise = Promise.resolve({ data: data.slice(0, 25), totalItems: data.length });
        props.api.get.mockReturnValueOnce(promise);
        props.api.get.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerList
                {...props}
                defaultSortColumn="column1"
            />,
        );

        expect.assertions(5);

        return promise.then(() => {
            expect(props.api.get).toHaveBeenCalledWith('/api/test', {
                pageSize: 25,
                pageNumber: 1,
                sortColumn: 'column1',
                sortDirection: 'ASC',
            });
            expect(wrapper.state()).toMatchObject({
                pageNumber: 1,
                totalPages: 3,
                sortColumn: 'column1',
                sortDirection: 'ASC',
            });
            expect(wrapper).toMatchSnapshot();

            wrapper.find(List).simulate('sortChange', { sortColumn: 'column1', sortDirection: 'DESC' });

            expect(props.api.get).toHaveBeenCalledWith('/api/test', {
                pageSize: 25,
                pageNumber: 1,
                sortColumn: 'column1',
                sortDirection: 'DESC',
            });
            expect(wrapper.state()).toMatchObject({
                sortColumn: 'column1',
                sortDirection: 'DESC',
            });
        });
    });

    test('renders errors correctly', () => {
        const promise = Promise.resolve({ status: 'error', errors: ['Some error'] });
        props.api.get.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerList
                {...props}
            />,
        );

        expect.assertions(3);

        return promise.then(() => {
            expect(wrapper.state()).toMatchObject({
                totalPages: 0,
            });
            expect(props.statusMessages.add).toHaveBeenCalledWith({ intent: 'danger', message: 'Some error' });
            expect(wrapper).toMatchSnapshot();
        });
    });
});
