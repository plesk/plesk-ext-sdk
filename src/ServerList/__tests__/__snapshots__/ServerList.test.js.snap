// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ServerList renders correctly 1`] = `
<div>
  <ListWrapper
    columns={
      [
        {
          "key": "column1",
          "title": "Column1",
        },
      ]
    }
    data={
      [
        {
          "column1": 0,
          "key": "0",
        },
        {
          "column1": 1,
          "key": "1",
        },
        {
          "column1": 2,
          "key": "2",
        },
        {
          "column1": 3,
          "key": "3",
        },
        {
          "column1": 4,
          "key": "4",
        },
      ]
    }
    onSortChange={[Function]}
    sortColumn={null}
    sortDirection="ASC"
  />
</div>
`;

exports[`ServerList renders errors correctly 1`] = `
<div>
  <ListWrapper
    columns={
      [
        {
          "key": "column1",
          "title": "Column1",
        },
      ]
    }
    data={[]}
    onSortChange={[Function]}
    sortColumn={null}
    sortDirection="ASC"
  />
</div>
`;

exports[`ServerList renders pagination correctly 1`] = `
<div>
  <ListWrapper
    columns={
      [
        {
          "key": "column1",
          "title": "Column1",
        },
      ]
    }
    data={
      [
        {
          "column1": 0,
          "key": "0",
        },
        {
          "column1": 1,
          "key": "1",
        },
        {
          "column1": 2,
          "key": "2",
        },
        {
          "column1": 3,
          "key": "3",
        },
        {
          "column1": 4,
          "key": "4",
        },
        {
          "column1": 5,
          "key": "5",
        },
        {
          "column1": 6,
          "key": "6",
        },
        {
          "column1": 7,
          "key": "7",
        },
        {
          "column1": 8,
          "key": "8",
        },
        {
          "column1": 9,
          "key": "9",
        },
        {
          "column1": 10,
          "key": "10",
        },
        {
          "column1": 11,
          "key": "11",
        },
        {
          "column1": 12,
          "key": "12",
        },
        {
          "column1": 13,
          "key": "13",
        },
        {
          "column1": 14,
          "key": "14",
        },
        {
          "column1": 15,
          "key": "15",
        },
        {
          "column1": 16,
          "key": "16",
        },
        {
          "column1": 17,
          "key": "17",
        },
        {
          "column1": 18,
          "key": "18",
        },
        {
          "column1": 19,
          "key": "19",
        },
        {
          "column1": 20,
          "key": "20",
        },
        {
          "column1": 21,
          "key": "21",
        },
        {
          "column1": 22,
          "key": "22",
        },
        {
          "column1": 23,
          "key": "23",
        },
        {
          "column1": 24,
          "key": "24",
        },
      ]
    }
    onSortChange={[Function]}
    sortColumn={null}
    sortDirection="ASC"
  />
  <Pagination
    current={1}
    onSelect={[Function]}
    total={3}
  />
</div>
`;

exports[`ServerList renders sort correctly 1`] = `
<div>
  <ListWrapper
    columns={
      [
        {
          "key": "column1",
          "title": "Column1",
        },
      ]
    }
    data={
      [
        {
          "column1": 0,
          "key": "0",
        },
        {
          "column1": 1,
          "key": "1",
        },
        {
          "column1": 2,
          "key": "2",
        },
        {
          "column1": 3,
          "key": "3",
        },
        {
          "column1": 4,
          "key": "4",
        },
        {
          "column1": 5,
          "key": "5",
        },
        {
          "column1": 6,
          "key": "6",
        },
        {
          "column1": 7,
          "key": "7",
        },
        {
          "column1": 8,
          "key": "8",
        },
        {
          "column1": 9,
          "key": "9",
        },
        {
          "column1": 10,
          "key": "10",
        },
        {
          "column1": 11,
          "key": "11",
        },
        {
          "column1": 12,
          "key": "12",
        },
        {
          "column1": 13,
          "key": "13",
        },
        {
          "column1": 14,
          "key": "14",
        },
        {
          "column1": 15,
          "key": "15",
        },
        {
          "column1": 16,
          "key": "16",
        },
        {
          "column1": 17,
          "key": "17",
        },
        {
          "column1": 18,
          "key": "18",
        },
        {
          "column1": 19,
          "key": "19",
        },
        {
          "column1": 20,
          "key": "20",
        },
        {
          "column1": 21,
          "key": "21",
        },
        {
          "column1": 22,
          "key": "22",
        },
        {
          "column1": 23,
          "key": "23",
        },
        {
          "column1": 24,
          "key": "24",
        },
      ]
    }
    onSortChange={[Function]}
    sortColumn="column1"
    sortDirection="ASC"
  />
  <Pagination
    current={1}
    onSelect={[Function]}
    total={3}
  />
</div>
`;
